### DISCUSSION TITLE
How to handle the data?

### DISCUSSION THREAD

**Discussion by fernandoruizarg** [eModbus/eModbus#260] [General]
Hi, I'm from Argentina and I have a question.... in the client example (RTU16example), the handling of the data is as follows:

`void handleData(ModbusMessage response, uint32_t token) `
 `{`
`  for (auto& byte : response) {`
`    Serial.printf("%02X ", byte);`
`  }`
`  Serial.println("");`
`}`

I have a keypad that outputs the following through the serial console, depending on which button has been pressed

`02 03 13 10 00 01 00 00 01 00 00 00`
  `|  |  |  |  |  |  |  |  |  |  |  |_Sixth button no press`
  `|  |  |  |  |  |  |  |  |  |  |____Fifth button no press`
  `|  |  |  |  |  |  |  |  |  |_______Fourth button no press` 
  `|  |  |  |  |  |  |  |  |__________Third button no press`
  `|  |  |  |  |  |  |  |_____________Second button no press`
  `|  |  |  |  |  |  |________________First button no press`
  `|  |  |  |  |  |___________________Last Communication data`
  `|  |  |  |  |______________________First Communication data`
  `|  |  |  |_________________________Last Address Data`
  `|  |  |____________________________First Address Data`
  `|  |_______________________________Command Code (Read the specified address data)`
  `|__________________________________Server ID`

How can I extract the data of each button pressed to a variable?

Thank you and I hope I have expressed myself correctly

**Comment by Miq1** [comment]
Hi, that should be simple, as all data is one byte only. The below is untested, but should give you the idea.
```
void handleButtons(ModbusMessage resp, uint32_t token) {
  uint8_t firstAddr, lastAddr, firstData, lastData, buttons[6];
  uint16_t offset = 2;
  offset = resp.get(offset, firstAddr);
  offset = resp.get(offset, lasAddr);
  offset = resp.get(offset, firstData);
  offset = resp.get(offset, lastData);
  for (uint8_t i = 0; i < 6; i++) {
    offset = resp.get(offset, buttons[i]);
  }
  // now do what you like with the data :)
  ...
}
```

**Reply by fernandoruizarg** [reply]
Hi Michael, thanks for your prompt reply.
I am really surprised by the work you have done on this library, it is really amazing and very helpful... thank you! ! !

Going back to my query, I'm a bit confused, what do I get in the offset variable? And what the resp.get() function returns...

Do I have any way to buy some more detailed documentation or guide for this library?

Thanks

**Comment by Miq1** [comment]
😀 
Obviously you did not consider reading the README... The very first line there says: 
> Read the docs at http://emodbus.github.io/!

There is a section called [Reading data from a message](https://emodbus.github.io/reading-from-a-modbusmessage#reading-data-from-a-message). Inside you will find your answers.

**Reply by fernandoruizarg** [reply]
Yes, I've read it, but some things haven't been clear to me... but I guess that's because of my limited programming knowledge :(

Thanks again

**Reply by Miq1** [reply]
If you need clarifications, just ask!

To answer your questions above in a global way, ``get()`` will start at the byte offset given as first argument to consume bytes to fill the variable given as the second. It the will put the converted bytes into the variable and return the updated value for the offset. That is why ``offset`` is taken both as first argument to ``get()`` and in turn takes the number given back from it. That way you may write ``get()`` after ``get()`` to fill your variables consecutively from the message.

**Reply by fernandoruizarg** [reply]
Understood !
And if the response received is variable (if I don't know how big it is), can I use resp.size() to find out what size to make the "for"?

**Reply by Miq1** [reply]
Yes. It may require some simple math, since ``size()`` will return the size in bytes overall, including the server ID and function code.

Your keypad is missing a length byte after these, by the way, a standard conforming message would have to have it for function code 0x03.

**Reply by fernandoruizarg** [reply]
Yes, it's true, return it

`[N] 112046| ESP32-MODBUS_v1.0.ino [  38] handleData: Response: serverID=3, FC=3, Token=00000001, length=10:`
`03 03 00 06 00 00 00 00 00 00`

> Yes. It may require some simple math, since size() will return the size in bytes overall, including the server ID and function code.

Then, the info of the buttons would be from offs = 4??

I mean, from resp[5] to resp[9] ?

**Reply by Miq1** [reply]
The message bytes are counted from 0, so the byte behind the ``06`` wouuld be offset 4.

**Comment by fernandoruizarg** [comment]
Another thing, this error is constant, it can happen after 8 seconds or 40 seconds... random.
What does it mean?

`[1;33m[E] 163646| ESP32-MODBUS_v1.0.ino [  80] handleError: Error response: E2 - CRC check error`
`[0m[1;33m[E] 196146| ESP32-MODBUS_v1.0.ino [  80] handleError: Error response: E2 - CRC check error`
`[0m[1;33m[E] 230746| ESP32-MODBUS_v1.0.ino [  80] handleError: Error response: E2 - CRC check error`
`[0m[1;33m[E] 242846| ESP32-MODBUS_v1.0.ino [  80] handleError: Error response: E2 - CRC check error`
`[0m[1;33m[E] 250946| ESP32-MODBUS_v1.0.ino [  80] handleError: Error response: E2 - CRC check error`
`[0m[1;33m[E] 257246| ESP32-MODBUS_v1.0.ino [  80] handleError: Error response: E2 - CRC check error`
`[0m[1;33m[E] 310646| ESP32-MODBUS_v1.0.ino [  80] handleError: Error response: E2 - CRC check error`
`[0m[1;33m[E] 342646| ESP32-MODBUS_v1.0.ino [  80] handleError: Error response: E2 - CRC check error`
`[0m[1;33m[E] 350046| ESP32-MODBUS_v1.0.ino [  80] handleError: Error response: E2 - CRC check error`
`[0m[1;33m[E] 368746| ESP32-MODBUS_v1.0.ino [  80] handleError: Error response: E2 - CRC check error`

**Reply by Miq1** [reply]
It hints to communications errors, most probably caused by partial messages. We have a long-lasting issue still open for RTU where messages are interrupted prematurely by a cause unknown. There are other possible causes, but it could be you are affected as well.

What is the ESP type you are using and how is the wiring between ESP and your keypad?

**Reply by fernandoruizarg** [reply]
Hello, sorry for the delay, I was on vacation.  The board is an ESP-32-NODEMCU, and the wiring is a 100 ohm shielded twisted pair cable 50 centimeters long

**Reply by Miq1** [reply]
That is a wire length that should not matter. What do you mean with the 100 Ohm here - is there a 100 Ohm resistor between both wires (should be 120 Ohm normally)? The most RS485 adapter boards I have seen had that already in place on the board; what is missing most of the time is a proper biasing on one end of the bus (pullup between A+ and Vcc and pulldown between B- and GND), but even that will not matter much with your short wires.

**Comment by fernandoruizarg** [comment]
other things, is there a way to generate a receive data without doing any polls? 
I mean, when I do an add.request, in backgrounds, an a request polls is generated, and when the server responds, the data will arrive in the callback function

`void handleData(ModbusMessage response, uint32_t token)`

but, if I had a server that every x amount of time, sends an I'm alive, is it possible to receive it on the client without any add.request() ???
Thanks a lot

**Reply by Miq1** [reply]
No, that is not possible due to the Modbus standards. A server may only send if requested and has to stay quiet the rest of the time.

**Comment by fernandoruizarg** [comment]
Yes, I know...
This is the scenario, since I send the request and receive the response, the time it takes is 47 milliseconds, in 35 keys on the bus, that will take 1.645 seconds, it is unacceptable to work. 
This keys has another setting, which does not require a request to respond, starts the communication and sends which key is pressed, with the same modbus frame...
So I thought, with this setting, the bus would be idle until each key is pressed, and the client would only have to wait to receive the data - handleData() - (It should be noted that on this bus, there would only be these keys, not other modbus server...)
Perhaps what you should do is not use the data as modbus, but use the first byte of the address, and from 5 to 10 for the buttons, and discard the rest of the frame, including the checksum.
I'm lost

**Reply by Miq1** [reply]
All that sounds not very efficient to me, to be honest.

So you got 35 keys that you need to react on in real-time manner. I suppose there will be only one or few keys being pressed at one time? Then polling is a bad idea, the more polling over Modbus RTU. Are these keys fixed to work over Modbus or do they have other means of communication? It is strange BTW that the keys shall be able to send their state as unrequested message - that is not Modbus RTU any more. 

What you would need is a "listen only" type of client, something completely non-standard and new. 

If you could neglect all functions a client normally has - why not set your keys in that send-unsolicited mode and have your ESP32 be another server on the bus, but as a listener only? It would get any message on the bus, but is unable to send itself.

With a second Serial interface and a second RS485 adapter you could even have a client running on the bus as well, but this will have to neglect all the errors that will be triggered by unsolicited messages coming in.

This may be doable, although there meight be pitfalls in it.

**Reply by fernandoruizarg** [reply]
as it is...
What remains for me to discover is if the keys have a multimaster implemented, that they wait for the idle bus to start the transmission and a collision does not occur, in that case, another server that listens to the keys could be the solution to what I am looking for
It is clear that the modbus protocol only works for devices that do not require real time, such as sensors or others...

**Reply by Miq1** [reply]
Depends on the speed - industrial Modbus is run with up to baud rates in the millions b/s. This is nothing the ESP32 can deliver, alas. I ran it closed-loop up to 800000 baud, but above of that the errors are avalanching.

I would suggest you give the listener a try, See [here](https://emodbus.github.io/modbusserver-rtu-api#void-registersniffermsrlistener-worker) for the docs.

**Reply by fernandoruizarg** [reply]
okay. Thank you

**Comment by fernandoruizarg** [comment]
Hello, a question, what are the IDs allowed to create a request? Could it be that FD or FF are not allowed?

Serial monitor:
> [0m[1;33m[E] 100| ESP32-MODBUS_v1.0.ino [ 208] loop: Error creating request: E1 - Invalid server

the loop portion:
` current_Millis = millis();`
` if (current_Millis - poll_Millis >= Interval_poll_time && nextPoll == true) {`
`    poll_Millis = current_Millis;`
`// Read register`
`    Error err = MB.addRequest(1, 0xFD, READ_HOLD_REGISTER, 0x100B, 0x0001);`
`    if (err!=SUCCESS) {`
`      ModbusError e(err);`
`      LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);`
`      nextPoll = false;`
`    }`
`  }`

**Reply by Miq1** [reply]
Right, the highest acceptable server ID is 247. You may download the Modbus standard document linked on the eModbus homepage to find this and other details about the standard.
